hw#5 Library application

jay-from-future opened the pull request:
hw#5 Library application

Contains following functionality:
- add items (Author, Genre, Book)
- remove items (Author, Genre, Book)

Commit:
Add Author and AuthorDao

Commit:
Switch to GenericDao, add Genre and Book Dao

Commit:
Refactoring: move DAO Exceptions into separate package

Commit:
Refactoring: remove test schema

Commit:
Add feature to show all books in library

Commit:
Add feature to show all authors in library

Commit:
Add feature to show all genres in library

Commit:
Add feature to add new author

Commit:
Test refactoring

Commit:
Adding functionality for items removal, adding book and genre

stvort commented on src/main/java/ru/otus/springlibrary/dao/mapper/BookRowMapper.java:
```diff
@@ -0,0 +1,40 @@
+        try {
+            Author author = authorDao.findById(rs.getInt("author_fk"));
+            Genre genre = genreDao.findById(rs.getInt("genre_fk"));
+            book = new Book(id, title, author, genre);
```
Лучше достать книгу(и) одним запросом переJOINив все таблицы-участники

jay-from-future replied:
Сделал через JOIN

stvort commented on src/main/java/ru/otus/springlibrary/domain/Author.java:
```diff
@@ -0,0 +1,35 @@
+@Getter
+public class Author {
+
+    private int id;
```
Лучше использовать long

jay-from-future replied:
Поправил, согласен. В схеме тоже поменял на bigint

stvort commented on src/test/java/ru/otus/springlibrary/dao/AuthorDaoImplTest.java:
```diff
@@ -0,0 +1,109 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class AuthorDaoImplTest extends BasicTest {
```
Дао Jdbc лучше тестировать через ```@JdbcTest + @Import нужного дао```

jay-from-future replied:
Исправил на 
```@JdbcTest + @ComponentScan``` (иначе Regular ```@Component``` beans are not loaded into the ApplicationContext, https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-testing.html)

Не уловил, для чего тут нужен ```@Import```?

stvort replied:
Чтобы только нужный дао попал в контекст. Замена ```@ComponentScan```

Commit:
Change id type from int to long

Commit:
Switch to use @JdbcTest for DAO tests

Commit:
Join data from different tables to retrieve all data for book in one query

stvort commented on src/main/java/ru/otus/springlibrary/dao/BookDaoImpl.java:
```diff
@@ -0,0 +1,70 @@
+public class BookDaoImpl implements BookDao {
+
+    private static final String SELECT_ALL_FROM_BOOK = "select b.id, title, firstname, lastname, genre from book b " +
+            "join author a on b.author_fk = a.id join genre g on b.genre_fk = g.id";
```
Тип JOIN лучше писать явно

jay-from-future replied:
Заменил на left outer join

stvort commented on src/test/java/ru/otus/springlibrary/dao/AuthorDaoImplTest.java:
```diff
@@ -0,0 +1,115 @@
+
+@RunWith(SpringRunner.class)
+@JdbcTest
+@ComponentScan
```
```@ComponentScan("ru.otus.springlibrary.dao")```

jay-from-future replied:
Спасибо, осознал. Теперь @Import({AuthorDaoImpl.class, GenericDaoImpl.class})

stvort commented on src/main/java/ru/otus/springlibrary/dao/GenericDaoImpl.java:
```diff
@@ -0,0 +1,80 @@
+@RequiredArgsConstructor
+public class GenericDaoImpl<T> implements GenericDao<T> {
+
+    private static final Logger logger = LoggerFactory.getLogger(GenreServiceImpl.class);
```
Точно GenreServiceImpl?

jay-from-future replied:
Поправил.

Commit:
Fix and refactor:
- replace ComponentScan with Import
- define join type explicitly
- fix Logger classes
